src/algebra/category/Group/injective.lean

```diff
@@ -105,4 +106,125 @@ module.Baer.injective $
       congr, }, },
 end
 
+namespace divisible_emb
+
+-- In this section we prove that any additive abelian group A can be embed into a divisible group,
+-- namely `∏ᵢ (ℚ⧸ℤ)` where `i` indexing over all morphism `A →+ ℚ⧸ℤ`
+
+local notation `I` := AddCommGroup.of A ⟶ AddCommGroup.of (rat_circle)
+local notation `Div` := AddCommGroup.of (Π (i : I), rat_circle.{u})
+
+noncomputable instance divisible_by_Div : divisible_by Div ℤ :=
+pi.divisible_by _
+
+-- We need to show this function is injective
+/--
+For every abelian group `A`, there is a group homomorphism `A ⟶ ∏ (i : A ⟶ ℚ/ℤ), ℚ/ℤ)`
+by evaluating: `a ↦ (f ↦ f a)`.
+-/
+@[simps] def to_Div : AddCommGroup.of A ⟶ Div :=
+{ to_fun := λ a i, i a,
+  map_zero' := by simpa only [map_zero],
+  map_add' := λ x y, by simpa only [map_add] }
+
+lemma to_Div_inj_of_exists
+  (h : ∀ (a : A), a ≠ 0 → ∃ (f : (⟨ℤ ∙ a⟩ : Module ℤ) ⟶ ⟨rat_circle⟩),
+    f ⟨a, submodule.subset_span $ set.mem_singleton _⟩ ≠ 0) :
+  function.injective $ to_Div A :=
+begin
+  contrapose! h,
+  simp only [function.injective, not_forall] at h,
+  obtain ⟨a, b, h1, h2⟩ := h,
+  refine ⟨a-b, λ r, h2 (sub_eq_zero.mp r), λ f, _⟩,
+  haveI injQZ : category_theory.injective (⟨rat_circle.{u}⟩ : Module ℤ) :=
+    injective_as_module_of_injective_as_Ab _,
+  let g : (⟨ℤ ∙ (a - b)⟩ : Module ℤ) ⟶ ⟨A⟩ := submodule.subtype (ℤ ∙ (a - b)),
+  haveI : mono g,
+  { rw Module.mono_iff_injective, apply submodule.injective_subtype, },
+  have f_eq := injective.comp_factor_thru f g,
+  rw [←fun_like.congr_fun f_eq ⟨a - b, submodule.subset_span (set.mem_singleton _)⟩,
+    comp_apply],
+  have := congr_fun h1 (injective.factor_thru f g).to_add_monoid_hom,
+  rw [to_Div_apply, to_Div_apply, ←sub_eq_zero, ←map_sub] at this,
+  convert this,
+end
+
+namespace infinite_order
+
+open submodule
+
+variables {A} {a : A}
+
+/--
+There is a morphism `⟨a⟩ ⟶ ℚ/ℤ` by `r • a ↦ a/2`.
+-/
+@[reducible] noncomputable def to_fun : (ℤ ∙ a) → rat_circle.{u} :=
+λ x, ulift.up $ quotient.mk' (rat.mk (coeff_of_span_singleton x) 2)
```

Comment:
I've been trying to define this in two steps, but not completely successful yet: first construct the isomorphism `zmod (add_order_of a) ≃+ (ℤ ∙ a)`, and then construct `zmod n →+ rat_circle` such that the image of 1 is nonzero (when n ≠ 1). There are certain defs you may use so that you don't have to prove all the lemmas below, like [add_monoid_hom.cod_restrict](https://leanprover-community.github.io/mathlib_docs/group_theory/submonoid/operations.html#add_monoid_hom.cod_restrict), [quotient_add_group.quotient_ker_equiv_of_surjective](https://leanprover-community.github.io/mathlib_docs/group_theory/quotient_group.html#quotient_add_group.quotient_ker_equiv_of_surjective), and [add_equiv.of_left_inverse'](https://leanprover-community.github.io/mathlib_docs/group_theory/submonoid/operations.html#add_equiv.of_left_inverse'). I think it's most convenient to use the first two; although the result would be `ℤ ⧸ zmultiples (add_order_of a) ≃+ zmultiples a` instead of involving `zmod` and `module.span`, it should still be easy to work with.